File: doc/haskell/LogicalFormsBehavior.hs
module LogicalFormsBehavior where

import Logic.LogicalForms
import GeometryLow.Orientation

shouldTestLogicalFormsBehavior :: Bool
shouldTestLogicalFormsBehavior = shouldOpsTerminatedListToDnf && shouldSpliceBackIfNeeded && shouldSpliceBack && shouldTakeSubTermTerminatedBy && shouldTakeSubTermTerminatedBy && shouldTakeSubTermsSeparatedBy



shouldOpsTerminatedListToDnf :: Bool
shouldOpsTerminatedListToDnf =
    opsTerminatedListToDnf Containment   [(1,And), (2,Or ), (3,And), (4,And)] == [[1,2,4], [1,3,4]] && opsTerminatedListToDnf Containment   [(3,And), (4,And), (1,And), (2,Or )] == [[2,4,1], [3,4,1]] &&
    opsTerminatedListToDnf Complementary [(1,Or ), (2,And), (3,Or ), (4,Or )] == [[1], [2,3], [4]]  && opsTerminatedListToDnf Complementary [(3,Or ), (4,Or ), (1,Or ), (2,And)] == [[2,3], [4], [1]] &&

    opsTerminatedListToDnf Containment   [('a', And), ('b', And), ('c', And), ('d', And)] == [['a', 'b', 'c', 'd']] &&
    opsTerminatedListToDnf Complementary [('a', Or ), ('b', Or ), ('c', Or ), ('d', Or )] == [['a'], ['b'], ['c'], ['d']] &&

    opsTerminatedListToDnf Containment   ([]::[OpTerminated Char]) == [[]] &&
    opsTerminatedListToDnf Complementary ([]::[OpTerminated Char]) == [] &&

    opsTerminatedListToDnf Containment   [('a', And), ('b', Or), ('c', Or), ('d', Or), ('e', And), ('f', And)] == [['a', 'b', 'f'], ['a', 'c', 'f'], ['a', 'd', 'f'], ['a', 'e', 'f']] &&
    opsTerminatedListToDnf Complementary [('a', And), ('b', Or), ('c', Or), ('d', Or), ('e', And), ('f', And)] == [['e', 'f', 'a', 'b'], ['c'], ['d']]



shouldSpliceBackIfNeeded :: Bool
shouldSpliceBackIfNeeded =
    spliceBackIfNeeded []                                 Nothing                == ([]::[[Char]]) &&
    spliceBackIfNeeded [] (Just [10, 20, 30])                                    == [[10, 20, 30]] &&
    spliceBackIfNeeded [['a', 'b', 'c'], ['x', 'y', 'z']] Nothing                == [                ['a', 'b', 'c'], ['x', 'y', 'z']] &&
    spliceBackIfNeeded [['a', 'b', 'c'], ['x', 'y', 'z']] (Just ['1', '2', '3']) == [['1', '2', '3',  'a', 'b', 'c'], ['x', 'y', 'z']]

shouldSpliceBack :: Bool
shouldSpliceBack =
    spliceBack []                                 [10, 20, 30]    == [[10, 20, 30]] &&
    spliceBack [['a', 'b', 'c'], ['x', 'y', 'z']] ['1', '2', '3'] == [['1', '2', '3',  'a', 'b', 'c'], ['x', 'y', 'z']]


shouldTakeSubTermTerminatedBy :: Bool
shouldTakeSubTermTerminatedBy =
    takeSubTermTerminatedBy And ('a', And) [('b', And), ('c', And), ('d', And)] == (['a']               , Just [('b', And), ('c', And), ('d', And)]) &&
    takeSubTermTerminatedBy And ('a', Or ) [('b', Or ), ('c', Or ), ('d', Or )] == (['a', 'b', 'c', 'd'], Nothing                                  ) &&
    takeSubTermTerminatedBy And ('a', Or ) [('b', Or ), ('c', And), ('d', Or )] == (['a', 'b', 'c']     , Just [('d', Or )                        ])


shouldTakeSubTermsSeparatedBy :: Bool
shouldTakeSubTermsSeparatedBy =
    takeSubTermsSeparatedBy And [('a', And), ('b', And), ('c', And), ('d', And)] == ([['a'], ['b'], ['c'], ['d']], Nothing                  ) &&
    takeSubTermsSeparatedBy And [('a', Or ), ('b', Or ), ('c', Or ), ('d', Or )] == ([                          ], Just ['a', 'b', 'c', 'd']) &&
    takeSubTermsSeparatedBy And [('a', Or ), ('b', Or ), ('c', And), ('d', Or )] == ([['a', 'b', 'c']           ], Just ['d'               ])

File: app/assets/math/haskell/spec/LogicalFormsBehavior.js
function shouldTour() {return vecEq(tour([]), []) && vecEq(tour([10]), [[10,10]]) && vecEq(tour([10,20]), [[10,20], [20,10]]) && vecEq(tour([10,20,30]), [[10,20], [20,30], [30,10]]);}

function shouldCycleToOpsTerminatedListWith()
{
	function posModR(a, m) {while (a < 0 || a >= m) a -= Math.sign(a) * m; return a;}
	function isConvexValue(a) {return posModR(a, 360) <= 180;}
	function changeIsConvexValue(a1, a2) {return isConvexValue(a2-a1);}
	function changeCountsAsConvexValue(referenceRotDir, a1, a2) {return isConvexValue(referenceRotDir == '+' ? a2-a1 : a1-a2);}
	function showAsDegree (a) {return ''+a+'°';}
	return true &&
		vecEq(cycleToOpsTerminatedListWith(changeIsConvexValue, showAsDegree, [0, 90, 180, 270]), [['0°', 'and'], ['90°', 'and'], ['180°', 'and'], ['270°', 'and']])
		true;
}

function shouldOpTerminate()
{
	function posModR(a, m) {while (a < 0 || a >= m) a -= Math.sign(a) * m; return a;}
	function isConvexValue(a) {return posModR(a, 360) <= 180;}
	function changeIsConvexValue(a1, a2) {return isConvexValue(a2-a1);}
	function changeCountsAsConvexValue(referenceRotDir, a1, a2) {return isConvexValue(referenceRotDir == '+' ? a2-a1 : a1-a2);}
	function showAsDegree (a) {return ''+a+'°';}
	return true &&
		vecEq(opTerminate(changeIsConvexValue, showAsDegree, 10 ,  20), ['10°', 'and']) &&
		vecEq(opTerminate(changeIsConvexValue, showAsDegree, 10 ,   0), ['10°', 'or' ]) &&
		vecEq(opTerminate(changeIsConvexValue, showAsDegree, 10,  220), ['10°', 'or' ]) &&
		true;
}

function shouldOpsTerminatedListToDnf()
{
	function posModR(a, m) {while (a < 0 || a >= m) a -= Math.sign(a) * m; return a;}
	function isConvexValue(a) {return posModR(a, 360) <= 180;}
	function changeIsConvexValue(a1, a2) {return isConvexValue(a2-a1);}
	function changeCountsAsConvexValue(referenceRotDir, a1, a2) {return isConvexValue(referenceRotDir == '+' ? a2-a1 : a1-a2);}
	function showAsDegree (a) {return ''+a+'°';}
	return true &&
		vecEq(opsTerminatedListToDnf('containment'  , [['a', 'and'], ['b', 'and'], ['c', 'and'], ['d', 'and']]), [['a', 'b', 'c', 'd']]) &&
		vecEq(opsTerminatedListToDnf('complementary', [['a', 'or' ], ['b', 'or' ], ['c', 'or' ], ['d', 'or' ]]), [['a'], ['b'], ['c'], ['d']]) &&
		vecEq(opsTerminatedListToDnf('containment'  , []), [[]]) &&
		vecEq(opsTerminatedListToDnf('complementary', []), []) &&
		vecEq(opsTerminatedListToDnf('containment'  , [['a', 'and'], ['b', 'or'], ['c', 'or'], ['d', 'or'], ['e', 'and'], ['f', 'and']]), [['a', 'b', 'f'], ['a', 'c', 'f'], ['a', 'd', 'f'], ['a', 'e', 'f']]) &&
		vecEq(opsTerminatedListToDnf('complementary', [['a', 'and'], ['b', 'or'], ['c', 'or'], ['d', 'or'], ['e', 'and'], ['f', 'and']]), [['e', 'f', 'a', 'b'], ['c'], ['d']]) &&
		true;
}

function shouldSpliceBackIfNeeded()
{
	return true &&
		vecEq(spliceBackIfNeeded([]                                , ['nothing'              ]), []                                              ) &&
		vecEq(spliceBackIfNeeded([]                                , ['just'   , [10, 20, 30]]), [[10, 20, 30]]                                  ) &&
		vecEq(spliceBackIfNeeded([['a', 'b', 'c'], ['x', 'y', 'z']], ['nothing'              ]), [              ['a', 'b', 'c'], ['x', 'y', 'z']]) &&
		vecEq(spliceBackIfNeeded([['a', 'b', 'c'], ['x', 'y', 'z']], ['just'   , [10, 20, 30]]), [[10, 20, 30,   'a', 'b', 'c'], ['x', 'y', 'z']]) &&
		true;
}


function shouldSpliceBack()
{
	return true &&
		vecEq(spliceBack([]                                , [10, 20, 30]), [[10, 20, 30]]                                ) &&
		vecEq(spliceBack([['a', 'b', 'c'], ['x', 'y', 'z']], [10, 20, 30]), [[10, 20, 30, 'a', 'b', 'c'], ['x', 'y', 'z']]) &&
		true;
}

function shouldTakeSubTermTerminatedBy()
{
	return true &&
		vecEq(takeSubTermTerminatedBy('and', ['a', 'and'], [['b', 'and'], ['c', 'and'], ['d', 'and']], 0), [['a']               , ['just', 0]]) &&
		vecEq(takeSubTermTerminatedBy('and', ['a', 'or' ], [['b', 'or' ], ['c', 'or' ], ['d', 'or' ]], 0), [['a', 'b', 'c', 'd'], ['nothing']]) &&
		vecEq(takeSubTermTerminatedBy('and', ['a', 'or' ], [['b', 'or' ], ['c', 'and'], ['d', 'or' ]], 0), [['a', 'b', 'c']     , ['just', 2]]) &&
		true;
}

function shouldTakeSubTermsSeparatedBy()
{
	return true &&
		vecEq(takeSubTermsSeparatedBy('and', [['a', 'and'], ['b', 'and'], ['c', 'and'], ['d', 'and']]), [[['a'], ['b'], ['c'], ['d']], ['nothing'                      ]]) &&
		vecEq(takeSubTermsSeparatedBy('and', [['a', 'or' ], ['b', 'or' ], ['c', 'or' ], ['d', 'or' ]]), [[                          ], ['just'   , ['a', 'b', 'c', 'd']]]) &&
		vecEq(takeSubTermsSeparatedBy('and', [['a', 'or' ], ['b', 'or' ], ['c', 'and'], ['d', 'or' ]]), [[['a', 'b', 'c']           ], ['just'   , ['d'               ]]]) &&
		true;
}
